runtime/lua/vim/shared.lua

```diff
@@ -10,6 +10,21 @@
 ---@diagnostic disable-next-line: lowercase-global
 vim = vim or {}
 
+--- Assert that a value has the expected type.
+---
+--- @generic T
+--- @param v T
+--- @param expected string Expected type
+local function expecttype(v, expected)
+  local actual = type(v)
+
+  -- Use an if statement with error() instead of assert() to avoid calling string.format if we don't
+  -- need it.
+  if actual ~= expected then
+    error(('expected %s, got %s'):format(expected, actual))
```

Comment:
Maybe worth including the actual value in addition to the types?

Reply:
For `nil` this shows `"expected string, got nil (nil)"` which is a little weird, but probably not weird enough to add a special case?

Reply:
Since this is the "fast path" (and the fail state), I'd agree it's not worth adding complexity to pretty this up.